File: packages/domain/src/provider-intelligence.ts
export const PROVIDER_INTEL_VERSION = "provider-intel-v1";

export type ProviderIntelType = "nursing_home" | "home_health" | "hospice";
export type DirectoryProjection = "CURRENT_DIRECTORY" | "KNOWN_NOT_CURRENT" | "EVIDENCE_ONLY";
export type Availability = "AVAILABLE" | "NOT_REPORTED" | "NOT_APPLICABLE" | "UNSUPPORTED";

export interface ProviderIntelParty {
  display_name: string;
  party_kind: "organization" | "individual";
  party_id: string;
  organization_id: string | null;
  relationship_type: string;
  raw_cms_role: string;
  ownership_percentage: number | null;
  temporal_status: "CURRENT" | "HISTORICAL" | "UNKNOWN";
  effective_from: string | null;
  confidence: string;
  person_publication_policy: string | null;
  public_profile: boolean;
}

export interface ProviderIntelChowEvent {
  event_id: string;
  effective_date: string | null;
  cms_raw_type_code: string | null;
  cms_raw_type_text: string | null;
  normalized_type: string | null;
  buyer_legal_entity: string | null;
  seller_legal_entity: string | null;
  source_dataset_key: string | null;
  source_dataset_id: string | null;
  confidence: string | null;
  safe_language: string | null;
  not_labeled_sale: boolean;
}

export interface ProviderIntelCmsStars {
  overall: number | null;
  health_inspection: number | null;
  staffing: number | null;
  quality_measure: number | null;
  label: string;
  not_trust_hub_rating: true;
  availability: string;
}

export interface NursingHomeProviderIntelligence {
  contract_version: string;
  canonical_id: string;
  provider_type: "nursing_home";
  identifier_type: "CCN";
  profile_intelligence_status: "READY" | "PARTIAL" | "EVIDENCE_ONLY" | "BLOCKED";
  directory: {
    official_status: string | null;
    projection: DirectoryProjection;
  };
  common: {
    display_name: string | null;
    legal_name: string | null;
    office: {
      address: string | null;
      city: string | null;
      state: string | null;
      zip: string | null;
      phone: string | null;
    };
  };
  nursing_home: {
    cms_stars?: ProviderIntelCmsStars;
    has_core_evidence?: boolean;
  } | null;
  quality_summary: {
    cms_stars: ProviderIntelCmsStars | null;
    nh_evidence_flags: {
      mds: boolean;
      pbj: boolean;
      fire: boolean;
      inspection: boolean;
      sff: boolean;
    } | null;
    synthetic_trust_hub_rating: false;
  };
  ownership_summary: {
    current_owners: ProviderIntelParty[];
    operators: ProviderIntelParty[];
    managers: ProviderIntelParty[];
    enrollment_organizations: ProviderIntelParty[];
    historical_ownership_observations: ProviderIntelParty[];
    unknown_ownership_observations: ProviderIntelParty[];
    counts: Record<string, number>;
    unresolved_edges_included: false;
  };
  chow: {
    ownership_change_history_available: boolean;
    confirmed_event_count?: number | null;
    events?: ProviderIntelChowEvent[] | null;
    reason?: string | null;
    zero_does_not_mean_no_change_occurred?: boolean;
  };
  evidence_as_of_by_family: Record<
    string,
    { band: string | null; source_modified_at: string | null; age_days: number | null }
  >;
  availability: Record<string, Availability | string>;
  limitations: string[];
  person_publication_policy: string;
  organization_public_route: false;
  fingerprint: string;
  profile_generated_at?: string;
}

export function isProviderIntelV1(value: unknown): value is NursingHomeProviderIntelligence {
  if (!value || typeof value !== "object") return false;
  const obj = value as NursingHomeProviderIntelligence;
  return (
    obj.contract_version === PROVIDER_INTEL_VERSION &&
    obj.provider_type === "nursing_home" &&
    obj.identifier_type === "CCN"
  );
}

export interface AgencyQualityMeasure {
  family: "hh_quality" | "hh_hhcahps" | "hospice_quality" | "hospice_cahps";
  measure_code: string;
  official_name: string;
  reporting_period: string | null;
  score: number | null;
  score_text: string | null;
  star_rating: number | null;
  availability: "REPORTED" | "SUPPRESSED" | "NOT_AVAILABLE" | "INSUFFICIENT_DATA";
  footnote: string | null;
}

export interface AgencyQualityFamily {
  family: AgencyQualityMeasure["family"];
  observation_count: number;
  by_availability: Record<string, number>;
  measures: AgencyQualityMeasure[];
}

export interface AgencyServiceOffering {
  code: string;
  official_field: string;
  offered: boolean | null;
}

interface AgencyIntelBase {
  contract_version: string;
  canonical_id: string;
  profile_intelligence_status: "READY" | "PARTIAL" | "EVIDENCE_ONLY" | "BLOCKED";
  directory: {
    official_status: string | null;
    projection: DirectoryProjection;
  };
  common: NursingHomeProviderIntelligence["common"];
  ownership_summary: NursingHomeProviderIntelligence["ownership_summary"];
  chow: {
    ownership_change_history_available: false;
    reason: "NO_PUBLIC_CMS_CHOW_SOURCE";
    confirmed_event_count: null;
    events: null;
    zero_does_not_mean_no_change_occurred: true;
  };
  geography: {
    coverage: {
      zip_observation_count: number;
      is_verified_county_service_area: false;
      is_verified_service_area: false;
    };
    county_service_area: "UNSUPPORTED";
  };
  evidence_as_of_by_family: NursingHomeProviderIntelligence["evidence_as_of_by_family"];
  availability: Record<string, Availability | string>;
  limitations: string[];
  person_publication_policy: string;
  organization_public_route: false;
  fingerprint: string;
  profile_generated_at?: string;
}

export interface CmsQualityOfPatientCareStar {
  value: number | null;
  footnote: string | null;
  label: "CMS Quality of Patient Care star";
  not_trust_hub_rating: true;
  availability: string;
}

export interface HomeHealthProviderIntelligence extends AgencyIntelBase {
  provider_type: "home_health";
  identifier_type: "HOME_HEALTH_CCN";
  home_health: {
    cms_quality_of_patient_care_star: CmsQualityOfPatientCareStar;
    ownership_type: string | null;
    has_core_evidence: boolean;
  } | null;
  quality_summary: {
    cms_quality_of_patient_care_star: CmsQualityOfPatientCareStar | null;
    families: AgencyQualityFamily[];
    synthetic_trust_hub_rating: false;
  };
  services: AgencyServiceOffering[];
}

export interface HospiceProviderIntelligence extends AgencyIntelBase {
  provider_type: "hospice";
  identifier_type: "HOSPICE_CCN";
  hospice: {
    ownership_type: string | null;
    office_county_name: string | null;
    office_county_is_not_service_area: true;
    has_core_evidence: boolean;
  } | null;
  quality_summary: {
    families: AgencyQualityFamily[];
    synthetic_trust_hub_rating: false;
  };
}

export function isHomeHealthIntelV1(value: unknown): value is HomeHealthProviderIntelligence {
  if (!value || typeof value !== "object") return false;
  const obj = value as HomeHealthProviderIntelligence;
  return (
    obj.contract_version === PROVIDER_INTEL_VERSION &&
    obj.provider_type === "home_health" &&
    obj.identifier_type === "HOME_HEALTH_CCN"
  );
}

export function isHospiceIntelV1(value: unknown): value is HospiceProviderIntelligence {
  if (!value || typeof value !== "object") return false;
  const obj = value as HospiceProviderIntelligence;
  return (
    obj.contract_version === PROVIDER_INTEL_VERSION &&
    obj.provider_type === "hospice" &&
    obj.identifier_type === "HOSPICE_CCN"
  );
}

export function cmsStarConsumerLabel(): string {
  return "CMS star rating";
}

export function directoryBanner(projection: DirectoryProjection): string | null {
  if (projection === "KNOWN_NOT_CURRENT") {
    return "This provider is known from CMS records but is not listed in the current CMS nursing-home directory. Absence from the current directory is not proof the facility stopped operating.";
  }
  return null;
}

export function chowAbsenceCopy(): string {
  return "No CMS ownership-change event is currently attached to this profile in the Trust Hub research graph. That is not proof that ownership never changed.";
}

export function chowUnsupportedCopy(providerType: "home_health" | "hospice"): string {
  const program = providerType === "home_health" ? "Home Health" : "Hospice";
  return `CMS does not publish a ${program} ownership-change event file. Absence of a CHOW record here is not proof that ownership never changed.`;
}

export function agencyDirectoryBanner(
  providerType: "home_health" | "hospice",
  projection: DirectoryProjection,
): string | null {
  if (projection === "KNOWN_NOT_CURRENT") {
    const directory =
      providerType === "home_health"
        ? "current CMS Home Health agency directory"
        : "current CMS Hospice General Information directory";
    return `This provider is known from CMS records but is not listed in the ${directory}. Absence from the current directory is not proof the provider stopped operating.`;
  }
  if (projection === "EVIDENCE_ONLY") {
    return "This CMS certification number has quality or other evidence on file but is not in the current Hospice General Information directory. That is not proof the provider stopped operating.";
  }
  return null;
}

export function cmsMeasureAvailabilityCopy(
  availability: AgencyQualityMeasure["availability"],
  score: number | null,
  scoreText: string | null,
): string {
  if (availability === "SUPPRESSED") return "Suppressed";
  if (availability === "INSUFFICIENT_DATA") return "Insufficient data";
  if (availability === "NOT_AVAILABLE") return "Not reported";
  if (scoreText?.trim()) return scoreText.trim();
  if (score == null) return "Not reported";
  return String(score);
}

export function partyCapCopy(shown: number, total: number): string | null {
  if (total <= shown) return null;
  return `Showing ${shown} of ${total} relationships from CMS/PECOS evidence.`;
}

export function currentOwnedByStatement(party: ProviderIntelParty): string | null {
  if (party.relationship_type !== "OWNED_BY" || party.temporal_status !== "CURRENT") return null;
  return `CMS/PECOS current ownership evidence lists ${party.display_name}.`;
}

File: services/ingest/src/care_ingest/provider_intelligence.py
"""Class-aware Provider Intelligence contract. No combined scores or fabricated CHOW."""

from __future__ import annotations

import hashlib
import json
from typing import Any

CONTRACT_VERSION = "provider-intel-v1"
TRANSFORMATION_VERSION = "provider-intel-v1"
NATIONAL_INTEL_SNAPSHOT = "senior-national-intel-v1"

IDENTIFIER_BY_TYPE = {
    "nursing_home": "CCN",
    "home_health": "HOME_HEALTH_CCN",
    "hospice": "HOSPICE_CCN",
}

DIRECTORY_PROJECTION = {
    "CURRENT_ACTIVE": "CURRENT_DIRECTORY",
    "ABSENT_FROM_CURRENT_DIRECTORY": "KNOWN_NOT_CURRENT",
}

PERSON_PUBLICATION_POLICY = "SOURCE_EVIDENCE_ONLY_NO_PUBLIC_PROFILE"

SAFE_CLAIMS = {
    "current_directory": "CMS reports this provider in its current directory.",
    "quality_available": "CMS quality data is available for this provider.",
    "chow_event": "CMS records show an ownership change effective {effective_date}.",
    "zip_coverage": "CMS ZIP coverage records include ZIP {zip_code}.",
}

UNSUPPORTED_CLAIMS = {
    "hh_chow_none": "This home-health agency has never changed ownership.",
    "hospice_chow_none": "This hospice has never changed ownership.",
    "county_service": "This provider serves this county.",
    "trust_rating": "Trust Hub rating",
    "combined_quality": "senior quality score",
}


def identifier_type_for(provider_type: str) -> str:
    if provider_type not in IDENTIFIER_BY_TYPE:
        raise ValueError(f"unsupported provider type: {provider_type}")
    return IDENTIFIER_BY_TYPE[provider_type]


def name_is_provider_identity() -> bool:
    return False


def provider_type_may_silently_change() -> bool:
    return False


def missing_quality_is_zero() -> bool:
    return False


def cms_star_is_trust_hub_rating() -> bool:
    return False


def zip_is_county_service() -> bool:
    return False


def chow_means_sale() -> bool:
    return False


def unknown_owner_is_former_owner() -> bool:
    return False


def combined_senior_denominator_allowed() -> bool:
    return False


def public_person_profile_allowed() -> bool:
    return False


def public_organization_route_allowed() -> bool:
    return False


def fingerprint_includes_generated_at() -> bool:
    return False


def project_directory_status(
    *,
    provider_type: str,
    official_status: str | None,
    has_class_directory_row: bool,
) -> str:
    if provider_type == "hospice" and not has_class_directory_row:
        return "EVIDENCE_ONLY"
    if official_status in DIRECTORY_PROJECTION:
        return DIRECTORY_PROJECTION[official_status]
    if has_class_directory_row:
        return "CURRENT_DIRECTORY"
    return "KNOWN_NOT_CURRENT"


def absence_means_closed() -> bool:
    return False


def chow_history_available(provider_type: str) -> bool | None:
    if provider_type == "nursing_home":
        return True
    if provider_type in {"home_health", "hospice"}:
        return False
    return None


def chow_unavailability_reason(provider_type: str) -> str | None:
    if provider_type in {"home_health", "hospice"}:
        return "NO_PUBLIC_CMS_CHOW_SOURCE"
    return None


def quality_value_state(raw: object | None, *, availability: str | None = None) -> str:
    if availability in {"SUPPRESSED", "NOT_AVAILABLE", "INSUFFICIENT_DATA"}:
        return availability
    if raw is None or raw == "":
        return "NOT_REPORTED"
    return "AVAILABLE"


def profile_intelligence_status(
    *,
    provider_type: str,
    directory_projection: str,
    has_identity: bool,
    has_core_evidence: bool,
) -> str:
    if not has_identity:
        return "BLOCKED"
    if directory_projection == "EVIDENCE_ONLY":
        return "EVIDENCE_ONLY"
    if directory_projection == "KNOWN_NOT_CURRENT":
        return "PARTIAL"
    if directory_projection == "CURRENT_DIRECTORY" and has_core_evidence:
        return "READY"
    return "PARTIAL"


def limitations_for(provider_type: str) -> list[str]:
    common = [
        "Ownership evidence is not a quality measure.",
        "Missing evidence is not a zero score.",
        "CMS ratings are CMS ratings, not Trust Hub ratings.",
    ]
    if provider_type == "nursing_home":
        return [
            *common,
            "UNKNOWN ownership does not prove an owner left.",
            "CHOW records are CMS transaction records, not Trust Hub judgments or sales.",
        ]
    if provider_type == "home_health":
        return [
            *common,
            "CMS ZIP coverage records are not a verified county service area.",
            "CMS does not publish a Home Health ownership-change event file.",
        ]
    return [
        *common,
        "The Hospice General Information directory is the current-provider denominator.",
        "Quality-only typed CCNs are not current-directory providers.",
        "CMS ZIP coverage records are not a verified county service area.",
        "CMS does not publish a Hospice ownership-change event file.",
    ]


def fingerprint_payload(obj: dict[str, Any]) -> str:
    body = {
        "canonical_id": obj.get("canonical_id"),
        "provider_type": obj.get("provider_type"),
        "directory": obj.get("directory"),
        "quality_summary": obj.get("quality_summary"),
        "ownership_summary": obj.get("ownership_summary"),
        "chow": obj.get("chow"),
        "geography": obj.get("geography"),
        "freshness": obj.get("evidence_as_of_by_family"),
        "limitations": obj.get("limitations"),
        "availability": obj.get("availability"),
    }
    return hashlib.sha256(json.dumps(body, sort_keys=True, default=str).encode()).hexdigest()
